### ISSUE TITLE
Pyright and mypy display an impossible case as being possible

### ISSUE THREAD

**Issue by Ovsyanka83** [microsoft/pyright#4642]
**Describe the bug**

**To Reproduce**
[Playground](https://pyright-playground.decorator-factory.su/?gzip=H4sIAIvj7GMC_0tJTVNIy8_XSEosslJIys_PUahRyMwrAZLFJUWaVlwKQJCZppCXX6KQWZyZV1ySmJecClKtA1atqZCYl4JVBsUEEChKLUtNzIkvqSwAq9LkAgBdmCJIewAAAA%3D%3D)

**Expected behavior**
`reveal_type` should display `str` 

**VS Code extension or command-line**
Both command line and VS code extension. Playground has the same behavior so it is enough to reproduce the error.

**Additional context**
Mypy has the same issue for some reason. I would assume that this scenario is not a bug but it is definitely a case where the typechecker lies to us (i.e. assert_never checks will fail).

Note that even though my example is quite abstract and simple, we had this situation in a production code base with many complex types, some of which intersected and revealed this behavior to us.

**Comment by erictraut** [comment]
The current behavior is as designed. The first `isinstance` type guard narrows the type of `bar` from `bool | int | str` to just `int | str`. The second `isinstance` type guard narrows the type from `int | str` to `bool | str` (since `bool` in the type system is a subtype of `int`.

I'm not sure what this code is attempting to do. If you want `str` to be the final result, you could reverse the two checks:

```python
if isinstance(bar, bool | str) and not isinstance(bar, bool): ...
```

In any case, I'm going to close this issue because pyright is working as intended in this case.
